amqp/transport.py

```diff
@@ -149,26 +149,43 @@ def having_timeout(self, timeout):
                     sock.settimeout(prev)
 
     def _connect(self, host, port, timeout):
-        entries = socket.getaddrinfo(
-            host, port, 0, socket.SOCK_STREAM, SOL_TCP,
-        )
-        for i, res in enumerate(entries):
-            af, socktype, proto, canonname, sa = res
+        e = None
+
+        # avoid additional DNS requests for AAAA if A succeeds
+        addr_types = (socket.AF_INET, socket.AF_INET6)
+        for n, family in enumerate(addr_types):
             try:
-                self.sock = socket.socket(af, socktype, proto)
+                entries = socket.getaddrinfo(
+                    host, port, family, socket.SOCK_STREAM, SOL_TCP)
+            except socket.gaierror:
+                # we may have depleted all our options
+                if n + 1 >= len(addr_types):
+                    # if getaddrinfo succeeded before for another address
+                    # family, reraise the previous socket.error since it's more
+                    # relevant to users
+                    raise (e
+                           if e is not None
+                           else socket.error("failed to connect"))
```

Comment:
Error message is not descriptive enough in my opinion. The message should at least mention that this is a DNS issue.

Reply:
Changed to "failed to resolve broker hostname"